Flex-model UI editor

joshuaunity opened the pull request:
## Description

This PR contains the UI modal dialogue to easily interact with an asset's flex model.

## Look & Feel

<img width="1153" height="1025" alt="image" src="https://github.com/user-attachments/assets/52d1d8cf-8226-4a2b-9b24-907ed95084fd" />

## How to test

1. Visit an asset's properties page @ http://localhost:5000/assets/<asset_id>/properties
2. Click on the `Edit Flex Model` Button
3. A modal pops up where you can add and edit flexmodel fields for the asset

## Further Improvements

None

## Related Items

This PR closes the issue #1423 

---

- [x] I agree to contribute to the project under Apache 2 License. 
- [x] To the best of my knowledge, the proposed patch is not based on code under GPL or other license that is incompatible with FlexMeasures

joshuaunity commented:
@nhoening @Flix6x

nhoening commented on flexmeasures/ui/views/assets/views.py:
```diff
@@ -331,6 +332,24 @@ def graphs(self, id: str, start_time=None, end_time=None):
         )
 
+    @login_required
+    @route("/<id>/flexmodel")
```
I don't think we should make its own page. A modal dialogue like the flex-context, which also does not have its own view.
I would place it on the properties view, see isse

joshuaunity replied:
i see, i will get to that when I've gotten this working properly

nhoening commented:
> @nhoening just like the flexContext, are each fields going to have a description as well?

Yes. Probably those can be gotten from the docs: https://flexmeasures.readthedocs.io/latest/features/scheduling.html#the-flex-models-corresponding-schedulers

nhoening commented on flexmeasures/ui/templates/base.html:
```diff
@@ -1032,7 +1032,123 @@
+        }
+
+        // Fetch Account Details
+        async function getAccount(accountId) {
```
We can use base.html to store a lot of JavaScript functions. It works.

We can also place them in ui/static/js/flexmeasures.js or better yet - to a more dedicated module in there, e.g. flex-context-and-model-dialogues.js (lines after 1091) and the three getters here could go on data-utils.js

Advantages:
- shorter base.html (easier reading)
- not all pages load these hundreds of lines
- maybe browsers can compress clean JavaScript files better

nhoening commented on flexmeasures/ui/templates/assets/asset_flexmodel.html:
```diff
@@ -0,0 +1,573 @@
+    // const typeThree = [String, Object];
+    const typeFour = [Array];
+
+    const sensorsApiUrl = `${apiBasePath}/api/v3_0/sensors?page=1&per_page=100&asset_id={{ asset.id }}`;
```
We upgraded the search for sensors to cover the whole site tree. The view tells us which asset the highest one is in the tree ("site asset"), and here in the JS we use that ID.

See [this PR](https://github.com/FlexMeasures/flexmeasures/pull/1582), you can basically copy that, it's just a few lines.

joshuaunity replied:
I see, so this will make sure every sensor, including the parent asset, is pulled.

joshuaunity replied:
but i dont want to pull from main into this branch to avoid an overwhelming amount of conflicts, the update will be needed to be made and resolved from the parent PR.

joshuaunity replied:
we can do that later

nhoening replied:
No need to pull main noe. The PR had a few lines for the context, which you can copy and use in this PR, but for the model.

nhoening replied:
No need to pull main noe. The PR had a few lines for the context, which you can copy and use in this PR, but for the model.

nhoening commented on flexmeasures/ui/templates/assets/asset_flexmodel.html:
```diff
@@ -0,0 +1,573 @@
+            }
+            assetFlexContext[fieldName] = "";
+            flexOptionsContainer.innerHTML = "";
+            await renderFlexContextForm(); // im rendring here so te card with the 'border-on-click' class is updated/exists
```
```suggestion
            await renderFlexContextForm(); //  rendering here so the card with the 'border-on-click' class is updated/exists
```

nhoening reviewed (changes requested):
Only a dry review, because I could not run the data migration yet.

Still hopefully giving you an iteration.

But I could not comment on the real functionality, as I would want to try it also.

joshuaunity commented on flexmeasures/ui/templates/assets/asset_flexmodel.html:
```diff
@@ -0,0 +1,573 @@
+
+            const requestBody = JSON.stringify({ flex_model: JSON.stringify(data) });
+
+            console.log("reqbody: ", requestBody)
```
yeah, i forgot to take it out when I was done testing

nhoening commented on flexmeasures/ui/templates/assets/asset_context.html:
```diff
@@ -293,7 +293,11 @@ <h5 class="modal-title pe-2">Edit Asset's FlexContext</h5>
 </script>
 
-<script>
+<script type="module" type="text/javascript">
```
What does `type="module"` do here? It is not really a module, I think.

joshuaunity commented:
> I believe the button should be inside the main body, not to the right. See placement of "Create sensor" in the box with the sensor table.

Do you want me to place it inside the `properties` container?

@nhoening

joshuaunity commented:
Sorry i accidentally deleted that comment 🤦

joshuaunity commented:
- [ ] When I select a field from the dropdown, the "About X" box should change content to explain that field.
- [ ] Loading the properties page, I see in the console "ReferenceError: can't access lexical declaration 'getFlexModel' before initialization", from the code as shown in the screenshot below
- [ ] There is still a line at the bottom of the tab
- [ ] In the value form, hitting Enter should set the value (like in flex context dialogue)
- [ ] How do I add new values to soc-gain when it already has one or two?
- [ ] I removed soc-gain and added a fixed value, I got this error: "Failed to update the flex model: {'soc-gain': ['Not a valid list.']} " (the value was added anyway ...)
- [ ] The "About Soc Gain " section should explain better that we can have several entries and how to add them. Other array fields, as well, of course.
- [ ] When I remove a field, it should show up again in the dropdown (like in the flex context dialogue)
- [ ] The info box about the whole dialogue should also link to the docs, like the flex-context info boy also does
- [ ] Can we also add the two green shades (for fixed versus sensor values) like in the flex context dialogue?

joshuaunity commented on flexmeasures/ui/templates/assets/asset_context.html:
```diff
@@ -288,12 +299,16 @@ <h5 class="modal-title pe-2">Edit Asset's FlexContext</h5>
     var marker = L
         .marker(
-            [{{ asset.latitude | replace("None", 10) }}, {{ asset.longitude | replace("None", 10) }}],
+            [{{ asset.latitude | replace("None", 10) }}, { { asset.longitude | replace("None", 10) } }],
```
I have my pages set to `HTML` and not `Django template`, this is probably the cause. Overall, this setting is very good for easily formatting lots of code at a time, but this is the only edge case. I do think this can be avoided if I properly wrap the Django variable in quotes

joshuaunity replied:
i only have "Jinja" on mine

joshuaunity commented:
@nhoening the efficiency fields have been all integrated

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1061 @@ <h3>All child assets for {{ asset.name }}
+            "roundtrip-efficiency": "Below 100%, this represents roundtrip losses (of charging & discharging), usually used for batteries. Can be percent or ratio [0,1] (defaults to 100%).",
+            "charging-efficiency": "Apply efficiency losses only at time of charging, not across roundtrip (defaults to 100%).",
+            "discharging-efficiency": "Apply efficiency losses only at time of discharging, not across roundtrip (defaults to 100%).",
+            "storage-efficiency": "This can encode losses over time, so each time step the energy is held longer leads to higher losses (defaults to 100%). Also read 11 about applying this value per time step across longer time spans.",
```
What is this "11" ? I guess you copied this from the docs and "11" was a link?

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1061 @@ <h3>All child assets for {{ asset.name }}
+            async function renderFlexInputOptions() {
+                const card = activeCard();
+                if (!card) {
+                    showToast(`Cant render input options without a selcted Field`, "info");
```
```suggestion
                    showToast(`Cant render input options without a selected field.`, "info");
```

nhoening replied:
I see this Toast just loading the properties page.

Also, I don't believe it has value for the user. They don't know what you mean with "render input options". Maybe print that to console.log and something more general to the Toast, if at all.

nhoening reviewed (changes requested):
Still happening (from last review): the `sensorsSearchResults` div is not disappearing now when I work on other field.

The new icon for field type is nice. I would actually move it into the info box fpr each field. Then users see it both when adding it but also when browsing the already-set fields.

Third: the info icon next to "Edit Asset's FlexModel" still needs content.

joshuaunity commented:
>The new icon for field type is nice. I would actually move it into the info box fpr each field. Then users see it both when adding it but also when browsing the already-set fields.

You mean i should add the icons inside the info container as well?

nhoening commented:
> > The new icon for field type is nice. I would actually move it into the info box fpr each field. Then users see it both when adding it but also when browsing the already-set fields.
> 
> You mean i should add the icons inside the info container as well?

No, I believe they could *only* live there.

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1077 @@ <h3>All child assets for {{ asset.name }}
+
+                if (isArray) {
+                    const plusIcon = document.createElement('i');
+                    plusIcon.className = 'fa fa-solid fa-plus ps-1';
```
Maybe add a `title` attribute that says what will happen, like "Add another setting"

nhoening replied:
Agreed, we should be consistent.

Users probably understand setting better than item. Item sounds more like a physical thing.

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1077 @@ <h3>All child assets for {{ asset.name }}
+            "soc-minima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form lower boundaries, e.g. to target a full car battery in the morning (defaults to NaN values).", "types": "Fixed value or Sensor." },
+            "soc-maxima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form upper boundaries at certain times (defaults to NaN values).", "types": "Fixed value or Sensor." },
+            "soc-targets": { "units": flexModelAllowedUnits.energy, "desc": "Exact set point(s) that the scheduler needs to realize (defaults to NaN values).", "types": "Fixed value or Sensor." },
+            "soc-gain": { "units": flexModelAllowedUnits.power, "desc": "SoC gain per time step, e.g. from a secondary energy source (defaults to zero). This field is an array whose values can be a string or a sensor.", "types": "An array of Fixed value or Sensor." },
```
"This field is an array" is not helping me as a user. 
I'd prefer a phrasing like "This field allows multiple settings."

Actually, I simply don't know how these settings combine - I guess they are additive (right, @Flix6x ?) and this is then important to tell the user.

So my proposal: "This field allows multiple settings, either fixed or dynamic, which add up to an aggregated gain."

(users won't also care to read about "string" versus "sensor" - we should use "fixed" versus "dynamic" instead)

The same change can be made at the soc-usage.

joshuaunity replied:
ok, i will make the update

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1077 @@ <h3>All child assets for {{ asset.name }}
+            "soc-max": { "units": flexModelAllowedUnits.energy, "desc": "A constant upper boundary for all values in the schedule (defaults to max soc target, if provided).", "types": "Fixed value or Sensor." },
+            "soc-minima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form lower boundaries, e.g. to target a full car battery in the morning (defaults to NaN values).", "types": "Fixed value or Sensor." },
+            "soc-maxima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form upper boundaries at certain times (defaults to NaN values).", "types": "Fixed value or Sensor." },
+            "soc-targets": { "units": flexModelAllowedUnits.energy, "desc": "Exact set point(s) that the scheduler needs to realize (defaults to NaN values).", "types": "Fixed value or Sensor." },
```
Can this say "One fixed value or a dynamic signal (via a sensor)." instead of "Fixed value or Sensor."?

I believe this explains better what we enable here.

(in the end, we should use these wordings on the flex context as well, in a later PR)

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1077 @@ <h3>All child assets for {{ asset.name }}
+                    fieldIconTitle = "A Fixed Value";
+                } else if (isArray) {
+                    fieldIconClass = iconMap["array"];
+                    fieldIconTitle = "An Array";
```
```suggestion
                    fieldIconTitle = "Multiple settings possible";
```

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1077 @@ <h3>All child assets for {{ asset.name }}
+                let fieldIconTitle = '';
+                if (isKeyValueObject) {
+                    fieldIconClass = iconMap["sensor"];
+                    fieldIconTitle = "A Sensor";
```
```suggestion
                    fieldIconTitle = "A dynamic signal (via a sensor)";
```

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1077 @@ <h3>All child assets for {{ asset.name }}
+                    fieldIconTitle = "A Sensor";
+                } else if (isBool) {
+                    fieldIconClass = iconMap["boolean"];
+                    fieldIconTitle = "A Boolean";
```
```suggestion
                    fieldIconTitle = "True or False (a boolean)";
```

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1077 @@ <h3>All child assets for {{ asset.name }}
+                }
+
+
+                iconSubContainer.appendChild(identityIcon);
```
I like the `identityIcon` - the only thing I'd change is that it is too close to a main action (`closeIcon`), obscuring it a bit.

Can it also appear in the bottom right (not top right), and also be grayed out a little?

nhoening reviewed (changes requested):
This seems functionally ready.

That is why my comments now are mostly about wording :)

Happy to merge this soon, from my end.

joshuaunity commented:
> You asked for a review but I see no commits...
> 
> Did you also find the cause of the other observation I had?

thats odd, let me cehck my local

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1078 @@ <h3>All child assets for {{ asset.name }}
+            "soc-minima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form lower boundaries, e.g. to target a full car battery in the morning (defaults to NaN values).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "soc-maxima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form upper boundaries at certain times (defaults to NaN values).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "soc-targets": { "units": flexModelAllowedUnits.energy, "desc": "Exact set point(s) that the scheduler needs to realize (defaults to NaN values).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "soc-gain": { "units": flexModelAllowedUnits.power, "desc": "SoC gain per time step, e.g. from a secondary energy source (defaults to zero). This field allows multiple settings, either fixed or dynamic, which add up to an aggregated gain.", "types": "An array of One fixed value or a dynamic signal (via a sensor)." },
```
```suggestion
            "soc-gain": { "units": flexModelAllowedUnits.power, "desc": "SoC gain per time step, e.g. from a secondary energy source (defaults to zero). This field allows multiple settings, either fixed or dynamic, which add up to an aggregated gain.", "types": Multiple settings possible - either fixed values or dynamic signals (via a sensor)." },
```

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1078 @@ <h3>All child assets for {{ asset.name }}
+            "soc-targets": { "units": flexModelAllowedUnits.energy, "desc": "Exact set point(s) that the scheduler needs to realize (defaults to NaN values).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "soc-gain": { "units": flexModelAllowedUnits.power, "desc": "SoC gain per time step, e.g. from a secondary energy source (defaults to zero). This field allows multiple settings, either fixed or dynamic, which add up to an aggregated gain.", "types": "An array of One fixed value or a dynamic signal (via a sensor)." },
+            "soc-usage": { "units": flexModelAllowedUnits.power, "desc": "SoC reduction per time step, e.g. from a load or heat sink (defaults to zero). This field allows multiple settings, either fixed or dynamic, which add up to an aggregated gain.", "types": "An array of One fixed value or a dynamic signal (via a sensor)." },
+            "roundtrip-efficiency": { "units": flexModelAllowedUnits.efficiency, "desc": "Below 100%, this represents roundtrip losses (of charging & discharging), usually used for batteries. Can be percent or ratio [0,1] (defaults to 100%).", "types": "Fixed value only." },
```
```suggestion
            "roundtrip-efficiency": { "units": flexModelAllowedUnits.efficiency, "desc": "Below 100%, this represents roundtrip losses (of charging & discharging), usually used for batteries. Can be percent or ratio [0,1]. Defaults to 100% (no loss).", "types": "Fixed value only." },
```

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1078 @@ <h3>All child assets for {{ asset.name }}
+            "soc-gain": { "units": flexModelAllowedUnits.power, "desc": "SoC gain per time step, e.g. from a secondary energy source (defaults to zero). This field allows multiple settings, either fixed or dynamic, which add up to an aggregated gain.", "types": "An array of One fixed value or a dynamic signal (via a sensor)." },
+            "soc-usage": { "units": flexModelAllowedUnits.power, "desc": "SoC reduction per time step, e.g. from a load or heat sink (defaults to zero). This field allows multiple settings, either fixed or dynamic, which add up to an aggregated gain.", "types": "An array of One fixed value or a dynamic signal (via a sensor)." },
+            "roundtrip-efficiency": { "units": flexModelAllowedUnits.efficiency, "desc": "Below 100%, this represents roundtrip losses (of charging & discharging), usually used for batteries. Can be percent or ratio [0,1] (defaults to 100%).", "types": "Fixed value only." },
+            "charging-efficiency": { "units": flexModelAllowedUnits.efficiency, "desc": "Apply efficiency losses only at time of charging, not across roundtrip (defaults to 100%).", "types": "Fixed value only." },
```
```suggestion
            "charging-efficiency": { "units": flexModelAllowedUnits.efficiency, "desc": "Apply efficiency losses only at time of charging, not across roundtrip. Defaults to 100% (no loss).", "types": "Fixed value only." },
```

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1078 @@ <h3>All child assets for {{ asset.name }}
+            "soc-usage": { "units": flexModelAllowedUnits.power, "desc": "SoC reduction per time step, e.g. from a load or heat sink (defaults to zero). This field allows multiple settings, either fixed or dynamic, which add up to an aggregated gain.", "types": "An array of One fixed value or a dynamic signal (via a sensor)." },
+            "roundtrip-efficiency": { "units": flexModelAllowedUnits.efficiency, "desc": "Below 100%, this represents roundtrip losses (of charging & discharging), usually used for batteries. Can be percent or ratio [0,1] (defaults to 100%).", "types": "Fixed value only." },
+            "charging-efficiency": { "units": flexModelAllowedUnits.efficiency, "desc": "Apply efficiency losses only at time of charging, not across roundtrip (defaults to 100%).", "types": "Fixed value only." },
+            "discharging-efficiency": { "units": flexModelAllowedUnits.efficiency, "desc": "Apply efficiency losses only at time of discharging, not across roundtrip (defaults to 100%).", "types": "Fixed value only." },
```
```suggestion
            "discharging-efficiency": { "units": flexModelAllowedUnits.efficiency, "desc": "Apply efficiency losses only at time of discharging, not across roundtrip. Defaults to 100% (no loss).", "types": "Fixed value only." },
```

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1078 @@ <h3>All child assets for {{ asset.name }}
+            "discharging-efficiency": { "units": flexModelAllowedUnits.efficiency, "desc": "Apply efficiency losses only at time of discharging, not across roundtrip (defaults to 100%).", "types": "Fixed value only." },
+            "storage-efficiency": {
+                "units": flexModelAllowedUnits.efficiency,
+                "desc": "This can encode losses over time, so each time step the energy is held longer leads to higher losses (defaults to 100%). The storage efficiency (e.g. 95% or 0.95) to use for the schedule is applied over each time step equal to the sensor resolution. For example, a storage efficiency of 95 percent per (absolute) day, for scheduling a 1-hour resolution sensor, should be pas ...
```
```suggestion
                "desc": "This can encode losses over time, so each time step the energy is held longer leads to higher losses. For example 95% (or 0.95) means a loss of 5% each step. Defaults to 100% (no loss). Note that the storage efficiency to use for the schedule is applied over each time step equal to the sensor resolution. For example, a storage efficiency of 95 percent  per (absolute) day, for scheduling a 1-hour resolution sensor, should be passed as a storage efficiency of 0.95<sup>1/24</sup> = 0.997865",
```

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1078 @@ <h3>All child assets for {{ asset.name }}
+            "storage-efficiency": {
+                "units": flexModelAllowedUnits.efficiency,
+                "desc": "This can encode losses over time, so each time step the energy is held longer leads to higher losses (defaults to 100%). The storage efficiency (e.g. 95% or 0.95) to use for the schedule is applied over each time step equal to the sensor resolution. For example, a storage efficiency of 95 percent per (absolute) day, for scheduling a 1-hour resolution sensor, should be pas ...
+                "types": "Fixed value only."
```
I believe this is wrong, as I can also choose a sensor.
```suggestion
                "types": "One fixed value or a dynamic signal (via a sensor)."
```

joshuaunity replied:
This is an oversight from me, i discussed about this field with @Flix6x and he said we keep it as a fixed value, but after testing with `roundtrip-effeciency` i forgot to set the other effeciency to Fixed value only.

But we can all quickly agree on this if we want to accept both or one.

joshuaunity replied:
>I set a sensor with kW unit for storage-efficiency. When the model is saved, I see:

I think this will also depend on the decison made in the comment

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1078 @@ <h3>All child assets for {{ asset.name }}
+                }
+
+                const identityIcon = document.createElement('i');
+                identityIcon.className = `${fieldIconClass} ps-1 fa-md float-end`;
```
Please change the color a bit, so it is not as prominent as the close (X) icon.

I suggest making it gray.

joshuaunity replied:
I just changed it now, can you confirm it fits what you envisioned

joshuaunity replied:
Yes, after checking now, I made all those changes in the other refactoring PR. i moved the code here now, though

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1078 @@ <h3>All child assets for {{ asset.name }}
+            "soc-max": { "units": flexModelAllowedUnits.energy, "desc": "A constant upper boundary for all values in the schedule (defaults to max soc target, if provided).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "soc-minima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form lower boundaries, e.g. to target a full car battery in the morning (defaults to NaN values).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "soc-maxima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form upper boundaries at certain times (defaults to NaN values).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "soc-targets": { "units": flexModelAllowedUnits.energy, "desc": "Exact set point(s) that the scheduler needs to realize (defaults to NaN values).", "types": "One fixed value or a dynamic signal (via a sensor)." },
```
```suggestion
            "soc-targets": { "units": flexModelAllowedUnits.energy, "desc": "Exact set point(s) that the scheduler needs to realize.", "types": "A sensor which records the set points as energy targets." },
```

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1078 @@ <h3>All child assets for {{ asset.name }}
+        const assetFlexModelFieldDesc = {
+            "soc-min": { "units": flexModelAllowedUnits.energy, "desc": "A constant lower boundary for all values in the schedule (defaults to 0).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "soc-max": { "units": flexModelAllowedUnits.energy, "desc": "A constant upper boundary for all values in the schedule (defaults to max soc target, if provided).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "soc-minima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form lower boundaries, e.g. to target a full car battery in the morning (defaults to NaN values).", "types": "One fixed value or a dynamic signal (via a sensor)." },
```
```suggestion
            "soc-minima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form lower boundaries, e.g. to target a full car battery in the morning", "types": "A sensor which records the set points as energy targets." },
```

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1078 @@ <h3>All child assets for {{ asset.name }}
+            "soc-min": { "units": flexModelAllowedUnits.energy, "desc": "A constant lower boundary for all values in the schedule (defaults to 0).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "soc-max": { "units": flexModelAllowedUnits.energy, "desc": "A constant upper boundary for all values in the schedule (defaults to max soc target, if provided).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "soc-minima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form lower boundaries, e.g. to target a full car battery in the morning (defaults to NaN values).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "soc-maxima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form upper boundaries at certain times (defaults to NaN values).", "types": "One fixed value or a dynamic signal (via a sensor)." },
```
```suggestion
            "soc-maxima": { "units": flexModelAllowedUnits.energy, "desc": "Set points that form upper boundaries at certain times.", "types": "A sensor which records the set points as energy targets." },
```

nhoening reviewed (changes requested):
It seems to work well now. Future improvements are obviously possible, but version 1 is there, I feel.

I suggest a number of improvements to docs, which is crucial for people to understand what we want them to do. Also, one mapox bug!

Plus, one observation, see below. Let me know how you want to tackle that.


I saw a small validation problem (which might be an issue for the model PR?).

I set a sensor with kW unit for storage-efficiency. When the model is saved, I see:

```
Failed to update the flex model: {'storage-efficiency': ['Must be greater than 0 and less than or equal to 1.']} 
```

This is a confusing validation error. Do you know how to improve this message, to be about the sensor unit? Probably something to uodate in the schema. 
Maybe this should be added to the model PR as a TODO ...

joshuaunity commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +258,1078 @@ <h3>All child assets for {{ asset.name }}
+                    $('[data-toggle="tooltip"]').tooltip();
+                });
+            });
+        addTileLayer(assetMap, '{ { mapboxAccessToken } }');
```
what errors did you see? so i can use that to check next time?

nhoening reviewed (changes requested):
Please add the graying out of the icon and (I suggest) move the deletion button back to the bottom.

joshuaunity commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -14,13 +14,34 @@
     <div class="row">
 
         <div class="col-md-2 on-top-md">
+            <div class="header-action-button">
```
I wasn't aware of this change, think this has been here for some time now. But i can move it.

nhoening commented on flexmeasures/data/schemas/scheduling/storage.py:
```diff
@@ -309,6 +350,68 @@ class DBStorageFlexModelSchema(Schema):
     )
 
+    soc_usage = fields.List(
+        VariableQuantityField("MW"),
```
I have a question here (and this applies to `soc_gain` as well (which is defined in the base branch for modelling) - state of charge is an energy measure in my opinion, and therefore energy units like kWh and MWh should be used.

Why are we using "MW" here - any reason you can recall?

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +259,1078 @@ <h3>All child assets for {{ asset.name }}
+            },
+            "prefer-charging-sooner": { "units": flexModelAllowedUnits.boolean, "desc": "Tie-breaking policy to apply if conditions are stable, which signals a preference to charge sooner rather than later (defaults to True). It also signals a preference to discharge later.", "types": "Boolean option only." },
+            "prefer-curtailing-later": { "units": flexModelAllowedUnits.boolean, "desc": "Tie-breaking policy to apply if conditions are stable, which signals a preference to curtail both consumption and production later, whichever is applicable (defaults to True).", "types": "Boolean option only." },
+            "power-capacity": { "units": flexModelAllowedUnits.power, "desc": "Device-level power constraint. How much power can be applied to this asset (defaults to the Sensor attribute capacity_in_mw).", "types": "One fixed value or a dynamic signal (via a sensor)." },
```
```suggestion
            "power-capacity": { "units": flexModelAllowedUnits.power, "desc": "Device-level power constraint. How much power can be applied to this asset.", "types": "One fixed value or a dynamic signal (via a sensor)." },
```

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +259,1078 @@ <h3>All child assets for {{ asset.name }}
+            "prefer-curtailing-later": { "units": flexModelAllowedUnits.boolean, "desc": "Tie-breaking policy to apply if conditions are stable, which signals a preference to curtail both consumption and production later, whichever is applicable (defaults to True).", "types": "Boolean option only." },
+            "power-capacity": { "units": flexModelAllowedUnits.power, "desc": "Device-level power constraint. How much power can be applied to this asset (defaults to the Sensor attribute capacity_in_mw).", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "consumption-capacity": { "units": flexModelAllowedUnits.power, "desc": "Device-level power constraint on consumption. How much power can be drawn by this asset.", "types": "One fixed value or a dynamic signal (via a sensor)." },
+            "production-capacity": { "units": flexModelAllowedUnits.power, "desc": "Device-level power constraint on production. How much power can be supplied by this asset. For PV curtailment, set this to reference your sensor containing PV power forecasts.", "types": "One fixed value or a dynamic signal (via a sensor)." },
```
```suggestion
            "production-capacity": { "units": flexModelAllowedUnits.power, "desc": "Device-level power constraint on production. How much power can be supplied by this asset. For PV curtailment (staying within promised output), set this to reference a sensor containing PV power forecasts.", "types": "One fixed value or a dynamic signal (via a sensor)." },
```

nhoening commented on flexmeasures/ui/templates/assets/asset_properties.html:
```diff
@@ -245,152 +259,1078 @@ <h3>All child assets for {{ asset.name }}
+                "desc": "This can encode losses over time, so each time step the energy is held longer leads to higher losses. For example 95% (or 0.95) means a loss of 5% each step. Defaults to 100% (no loss). Note that the storage efficiency to use for the schedule is applied over each time step equal to the sensor resolution. For example, a storage efficiency of 95 percent  per (absolute) day, ...
+                "types": "Fixed value only."
+            },
+            "prefer-charging-sooner": { "units": flexModelAllowedUnits.boolean, "desc": "Tie-breaking policy to apply if conditions are stable, which signals a preference to charge sooner rather than later (defaults to True). It also signals a preference to discharge later.", "types": "Boolean option only." },
```
Actually, the UI defaults this to `False`, not `True` - if I add the field, then just confirm with the set value button, I have selected this to `False`. I found this a little confusing. Maybe we can make this (and `prefer-curtailing-later`) be set to `True` per default?

nhoening reviewed (changes requested):
I didn't encounter new errors, just finding new small ways to make users have a less hard time with this complex issue.

Most of my comments can just be applied directly, but two might require some simple action.
